### ISSUE TITLE
API/namespace: move a dir into empty other dir overwrites it

### ISSUE THREAD

**Issue by onnozweers** [dCache/dcache#5331]
Dear dCache devs,

When we move a directory with the /namespace API call into another dir, and that target dir is empty, then the moved dir overwrites the target dir.

When on the other hand the target dir contains a file, then the source dir does _not_ overwrite the target dir, but the call returns a "400 Bad Request".

Here's the call:

```
curl <auth headers> -H 'accept: application/json' --fail --silent --show-error -H 'content-type: application/json' -X POST https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace//users/onno/disk/dir1 -d '{"action":"mv","destination":"/users/onno/disk/dir2/"}'
{"status":"success"}
```

This seems to be a bit inconsistent. Is this expected behaviour?

Cheers,
Onno

**Comment by paulmillar** [comment]
Hi Onno,

I think this behaviour is expected, but perhaps is isn't properly document.

With the `mv` command, dCache takes the `destination` field as the new name of the source.  So, using your example, if the POST command targeting `/api/v1/namespace/users/onno/disk/dir1` with entity:

    {"action":"mv","destination":"/users/onno/disk/dir2/"}

were to be successful, then the contents of `/users/onno/disk/dir1` would be available under `/users/onno/disk/dir2`.  In other words, after a successful mv operation, the file `/users/onno/disk/dir1/test-data.h5` will be available as the file `/users/onno/disk/dir2/test-data.h5`.

If the target is an empty directory then dCache will delete the directory automatically, so allowing the `mv` operation to succeed.  However, if `/users/onno/disk/dir2/` is non-empty then it cannot be deleted and the operation fails.

To move the directory `dir1` into `dir2` the command would look like:

    curl [...] -X POST https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace/users/onno/disk/dir1 \
          -d '{"action":"mv","destination":"/users/onno/disk/dir2/dir1"}'


Cheers,
Paul.

**Comment by onnozweers** [comment]
Hi Paul,

Still, I'd find it less confusing if moving a dir on to another (empty) dir would return an error. The target directory may be empty, but it may contain important attributes like disk/tape policy, a space token, ACLs, besides ownership and posix permissions. Would these attributes be overwritten?

Cheers,
Onno

**Comment by paulmillar** [comment]
Hi Onno,

To answer your direct question: yes, if a rename targets an existing, empty directory so that target directory is deleted then the target directory's attributes/tags/ACLs/POSIX permissions are all lost when the directory is deleted.

Looking into this some more, dCache will behave in the same way when renaming directories through dcap, FTP and NFS.  Those protocols support renaming where the target directory is deleted by the move operation, but only if the target directory is empty.  Otherwise, the operation fails.

In the case of NFS, this is exactly how it is supposed to work (v3, v4.1):

https://tools.ietf.org/html/rfc1813#section-3.3.14
https://tools.ietf.org/html/rfc5661#page-473

The semantics for FTP are unclear.  RFC 3659 gives an example where a file is overwritten during a rename, but doesn't mention renaming directories.

https://tools.ietf.org/html/rfc3659

For WebDAV, there is a flag that describes whether the target should be overwritten.  If overwrite is enabled, then the target is deleted (recursively!) before the move operation is performed.

https://tools.ietf.org/html/rfc4918#section-9.9

Other protocols (SRM, xrootd) have the semantics you describe: they allow renaming but only if the destination doesn't already exist: SRM make this clear, while xrootd the expected behaviour seems to be undocumented.  Using either protocol, the client must first remove the target before rename operation.

https://xrootd.slac.stanford.edu/doc/dev50/XRdv400.htm#_Toc30176901
https://sdm.lbl.gov/srm-wg/doc/SRM.v2.2.html#_Toc241633107

So, in summary, the frontend behaviour is consistent with dcap, FTP and NFS, where NFS semantics are quite well-defined.  WebDAV is more "trigger happy" and will overwrite non-empty directories, while xrootd and SRM are more conservative and do not overwrite when renaming.

I think the most important point is that the behaviour should be well documented (see #5338).  With the current semantics, a client can check whether the target exists before issuing the rename operation.  This is overhead (and non-atomic) so if there's a strong use-case for renaming without overwriting, we could support this by adding a flag to control whether or not the mv action should overwrite the destination.

Cheers,
Paul.

**Comment by onnozweers** [comment]
Hi Paul,

Wow, you really studies this! I had no idea that this behaviour was already quite common. In this case, I fully understand your choice.

We don't have such a "strong use-case", so you can save yourself the effort. Instead, we'll do an extra check on the client side; that should be good enough and easy to implement. Thanks for looking into this!

Cheers,
Onno
